File: DiaryNutrition/app/src/androidTest/java/ediger/diarynutrition/db/WaterDaoTest.java
package ediger.diarynutrition.db;


import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.List;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.room.Room;
import androidx.test.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;
import ediger.diarynutrition.LiveDataTestUtil;
import ediger.diarynutrition.data.DiaryDatabase;
import ediger.diarynutrition.data.Water;
import ediger.diarynutrition.data.dao.WaterDao;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;

@RunWith(AndroidJUnit4.class)
public class WaterDaoTest {

    static final Water WATER = new Water(250, 111111);

    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    private DiaryDatabase mDatabase;
    private WaterDao mWaterDao;

    @Before
    public void initDb() {
        mDatabase = Room.inMemoryDatabaseBuilder(InstrumentationRegistry.getContext(),
                DiaryDatabase.class)
                .allowMainThreadQueries()
                .build();
        mWaterDao = mDatabase.waterDao();
    }

    @After
    public void closeDb() {
        mDatabase.close();
    }

    @Test
    public void insertAndDeleteById() throws Exception {
        mWaterDao.insertWater(WATER);
        mWaterDao.deleteWaterById(1);

        List<Water> waterList = LiveDataTestUtil.getValue(mWaterDao.loadWater(0, 200000));
        assertThat(waterList.size(), is(0));
    }

    @Test
    public void insertAndGetAll() throws Exception {
        mWaterDao.insertWater(WATER);
        mWaterDao.insertWater(new Water(100, 111111));

        List<Water> waterList = LiveDataTestUtil.getValue(mWaterDao.loadWater(0, 200000));
        assertThat(waterList.size(), is(2));
        assertWater(waterList.get(0), 1, WATER.getAmount(), WATER.getDatetime());
    }

    @Test
    public void insertAndGetAmountSum() throws Exception {
        mWaterDao.insertWater(WATER);
        mWaterDao.insertWater(new Water(100, 111111));

        Water sumWater = LiveDataTestUtil.getValue(mWaterDao.getWaterSum(0, 200000));
        assertThat(sumWater, notNullValue());
        assertThat(sumWater.getAmount(), is(350));
    }

    @Test
    public void getEmptyAmountSum() throws Exception {
        Water sumWater = LiveDataTestUtil.getValue(mWaterDao.getWaterSum(0, 200000));

        assertThat(sumWater, notNullValue());
        assertThat(sumWater.getAmount(), is(0));
    }

    private void assertWater(Water water, int id, int amount, long datetime) {
        assertThat(water, notNullValue());
        assertThat(water.getId(), is(id));
        assertThat(water.getAmount(), is(amount));
        assertThat(water.getDatetime(), is(datetime));
    }

}

File: DiaryNutrition/app/src/androidTest/java/ediger/diarynutrition/db/WaterDaoTest.kt
package ediger.diarynutrition.db

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import ediger.diarynutrition.data.source.DiaryDatabase
import ediger.diarynutrition.data.source.entities.Water
import ediger.diarynutrition.getOrAwaitValue
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.Is.`is`
import org.hamcrest.core.IsNull.notNullValue
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

fun Water.assertWater(original: Water) {
    assertThat(this, notNullValue())
    assertThat(this.amount, `is`(original.amount))
    assertThat(this.datetime, `is`(original.datetime))
}

@ExperimentalCoroutinesApi
@RunWith(AndroidJUnit4::class)
class WaterDaoTest {

    @get:Rule
    var instantExecutorRule = InstantTaskExecutorRule()

    private lateinit var database: DiaryDatabase

    @Before
    fun initDb() = runBlockingTest {
        database = Room.inMemoryDatabaseBuilder(
                ApplicationProvider.getApplicationContext(),
                DiaryDatabase::class.java
        ).build()
    }

    @After
    fun closeDb() {
        database.close()
    }

    @Test
    fun populateAndGetRecords() = runBlockingTest {
        val water = Water(1000, 222222)
        database.waterDao().populateWater(listOf(WATER, water))
        val waterList = database.waterDao().getWaterList(0, 300000).getOrAwaitValue()
        assertThat(waterList.size, `is`(2))

        waterList.find { it.datetime == WATER.datetime }!!.assertWater(WATER)
        waterList.find { it.datetime == water.datetime }!!.assertWater(water)
    }

    @Test
    fun insertAndDeleteById() = runBlockingTest {
        database.waterDao().insertWater(WATER)

        var waterList = database.waterDao().getWater()
        assertThat(waterList.size, `is`(1))

        database.waterDao().deleteWaterById(1)

        waterList = database.waterDao().getWater()
        assertThat(waterList.size, `is`(0))
    }

    @Test
    fun insertAndGetAmountSum() = runBlockingTest {
        val water = Water(1000, 222222)
        database.waterDao().populateWater(listOf(WATER, water))

        val waterSum = database.waterDao().getWaterSum(0, 300000).getOrAwaitValue()
        assertThat(waterSum, notNullValue())
        assertThat(waterSum.amount, `is`(WATER.amount + water.amount))
    }

    @Test
    fun getAmountSum_empty() {
        val waterSum = database.waterDao().getWaterSum(0, 300000).getOrAwaitValue()

        assertThat(waterSum, notNullValue())
        assertThat(waterSum.amount, `is`(0))
    }

    companion object {
        val WATER = Water(250, 111111)
    }
 }
